Tree Checkbox Filtering

alexmassen-hane opened the pull request:
Changed the website to have a hierarchical checkboxes for the regions and subregions.
Will need to check with Jamie and or Cameron for text styling and font sizing. Unit tests are still required. 

![Screen Shot 2022-10-04 at 6 00 33 pm](https://user-images.githubusercontent.com/104514872/193792023-b8c5378f-72c6-49dc-b78e-8b98c30641c9.png)

jdddog commented on components/filter/SubregionForm.tsx:
```diff
@@ -63,19 +63,24 @@ const SubregionCheckbox = ({ region, subregion }: SubregionCheckboxProps) => {
+    <HStack>
+      <Box width="15%" />
+      <Box width="85%">
+        <VStack width="inherit" align="left">
+          {regions[region].map((subregion): ReactElement => {
+            return <SubregionCheckbox key={subregion} region={region} subregion={subregion} />;
+          })}
+        </VStack>
```
You can remove the HStack and the two Box elements, then just set the left padding or margin (e.g. pl="10px") on VStack.

jdddog reviewed (changes requested):
Nice work Alex, it works pretty well! 

Here is some feedback on some fine tuning to the interaction logic and the aesthetic of the design:

* The accordion title can just be Region.
* The arrow to the right of the region should point right rather than left by default.
* When a subset of subregions are checked, then the region checkbox should show an indeterminate state like below for parent checkbox.
![image](https://user-images.githubusercontent.com/5715104/194789720-1d19a50a-b70f-43e8-80b3-5c728730a3a5.png)
* When the parent checkbox is selected, the children should open. This includes clicking the parent name.
* In terms of the aesthetics of the hierarchical checkboxes, we should differentiate them from the styling used for the top level accordions used for region, open access, institution type etc.
  * You can make a custom theme variant for the accordion to accomplish this. You can just copy paste the filterForm theme, rename it and modify it.
  * The main thing is to remove the styling for parent accordion header (the white box, and the hover effects), so that this has the same background colour as the children and make the parent the same height as the child. Also use the same font size for parent and child text.
  *  Set a fixed width in pixels for the child indent (there is a comment below on how to do this).

Make sure to keep this current behaviour:
* Clicking the parent name should select / unselect the parent checkbox.
* Clicking the parent arrow should just open and close the list.

alexmassen-hane commented:
I have made the recursive general approach to the region tree filtering, but it may not be necessary to implement if the previous version just needs changes. 

The recursive version has a logic bug in where goes into an infinite loop and breaks the site if the parent checkbox is checked. The part that breaks the code is line 231 of RegionRecursiveTreeFiltering.tsx, but it is left commented out.

I have concerns that this may not be the best approach for generalised filtering. 

![image](https://user-images.githubusercontent.com/104514872/194797904-2ad27644-6bb7-465d-8c2e-7a09ca5766d6.png)

alexmassen-hane commented:
- Changed name to just "Region"
- Changed to Right Chevron for expand arrow.
- Added intermediate marked checkbox for parent.
- Added ability to open parent accordion by marking parent as checked, but will not open it again if user unchecks the checkbox and it is closed. 
- Forced styling to remove the parent hover-over change colour.
- Made parent and child accordion item panel share the same colour. 
- Made checkbox indent based on the level of the tree the checkbox is in using padding left. 
- Mild performance improvements. Can be problematic with the recursive nature of this new checkbox system.
- Added e2e unit tests for filtering to resolve specific country and institution using the filtering checkbox. 

![image](https://user-images.githubusercontent.com/104514872/196365343-2cc9eca7-75f4-4880-b9f4-68d5538c9724.png)

alexmassen-hane commented:
- Moved styling to theme file accordion.ts and removed boarders for checkboxes. 
- Removed chakra class css names from tests and introduced "componentTestId" string to FilterAccordionItem and AccordionButton to identify buttons for unit tests. 
- Moved tree object for region filtering to FilterForm.
- Moved hierarchal checkbox function to common and made it independent of regions/subregions.  
- Need have unit tests and functionality for mobile devices reviewed. 
- Need feedback from users for the new checkbox system and see how they go about using it.

![image](https://user-images.githubusercontent.com/104514872/197116598-3a059afa-62e4-41fc-91f4-3779996793ec.png)

alexmassen-hane commented:
- Increased padding of the last accordion item.
- Fixed unit tests for Safari mobile. All tests should now pass.
- Created theme file for Text component to neaten up code.
- Removed ability to check the checkbox by clicking on the name. Now users can only click on the checkbox itself to check them.
- When parent checkbox is checked, the panel still opens and marks all child checkboxes. 

![image](https://user-images.githubusercontent.com/104514872/197999630-a22b490d-7fd7-433f-afe0-42c5f4b89858.png)

jdddog commented on components/common/HierarchicalCheckbox.tsx:
```diff
@@ -0,0 +1,213 @@
+  );
+};
+
+export interface RecursiveTreeCheckboxProps {
+  level: number;
+  type: string;
+  text: string;
+  children?: Array<RecursiveTreeCheckboxProps>;
+}
+
+// Recursively go down branches and render checkboxes and input values into Formik
+const RecursiveTreeCheckbox = (branch: RecursiveTreeCheckboxProps): ReactElement => {
+  if (branch.children === undefined) {
+    return <></>;
+  } else {
```
With the above suggestions regarding recursion / theming, we could simplify the RecursiveTreeCheckboxProps type,  e.g. level isn't needed anymore. We won't have to add the if(branch.children) checks with these type definitions too.

Example types:
```typescript
export interface HierarchicalCheckboxParent {
  type: string;
  text: string;
  children: Array<HierarchicalCheckboxChild>;
}

export interface HierarchicalCheckboxChild {
  type: string;
  text: string;
}
```

Example config:
```typescript
export const regionsConfig: Array<HierarchicalCheckboxParent> = [
  {
    type: "region",
    text: "Africa",
    children: [
      {
        type: "subregion",
        text: "Northern Africa",
      },
      {
        type: "subregion",
        text: "Sub-Saharan Africa",
      },
    ],
  },
  {
    type: "region",
    text: "Americas",
    children: [
      {
        type: "subregion",
        text: "Latin America and the Caribbean",
      },
      {
        type: "subregion",
        text: "Northern America",
      },
    ],
  },
];
```

jdddog reviewed (changes requested):
Hey Alex, it is looking good and working really well. See below for some suggestions about the theme and changes to the code.

It would also be good, if for the children (not the parents), we made it so that clicking the label checks / unchecks the checkboxes.

This is what the styling suggestions make the layout look like:

![image](https://user-images.githubusercontent.com/5715104/198929233-48f571f9-7cb3-4631-aa53-f8401160173c.png)

alexmassen-hane commented on components/common/HierarchicalCheckbox.tsx:
```diff
@@ -0,0 +1,213 @@
+          return (
+            <AccordionItem
+              key={branchSub.text}
+              pl={`${34 * (branchSub.level - 1)}` + "px"}
+              _last={{ paddingBottom: "8px" }}
```
If I uncomment those two lines and use your suggestion for the theme, it breaks the indent look of the checkboxes on my browsers. We need a style that increases the padding on each branch.level

![image](https://user-images.githubusercontent.com/104514872/199436614-97817f6e-d7bb-4303-be75-f0613d89510d.png)

jdddog reviewed (approved):
Nice work Alex. I've just made a few tweaks to the styling and some adjustments to how the components of the tree checkbox are defined.
